Refactor and Modularize Code for Flexibility and Config-driven Setup

shantanuparab-tr opened the pull request:
This PR includes comprehensive refactoring to enhance modularity, flexibility, and configuration-driven behavior. Key changes are as follows:

- **Dynamic Robot Configuration**: Replaced hardcoded robot instances with a dynamic dictionary of robots (`self.robots`), initialized based on YAML config values for flexible setup of leader and follower robots.
- **Camera Configuration**: Updated camera setup to read from configuration, allowing any number of cameras to be defined in YAML and removing reliance on hardcoded camera names.
- **Gripper and Joint Operations**: Refactored gripper control and joint positioning to operate based on dictionary values, iterating over available robots rather than assuming specific follower/leader pairs.
- **Base Setup and Dynamic Import**: Added conditional imports and setup for mobile base (SLATE), using config-driven toggling of base features and dynamic import for related classes.
- **Task Configuration**: Extracted task details from YAML for dynamic handling of dataset directories and episode lengths, allowing for task-specific adjustments.

## Additional Notes
- Replaced instances of hardcoded values and duplicated code blocks with modular methods that adapt based on the configuration.
- This refactoring improves maintainability, simplifies scaling to new robots or cameras, and enables config-only modifications for different setups.
  
## Testing
- Verified initialization and control of leader/follower robots and grippers.
- Tested camera subscription and image processing with various camera configurations.
- Checked conditional base setup with both mobile and stationary configurations.

shantanuparab-tr committed:
Use YAML file for Bringup,Teleop and Sleep

shantanuparab-tr committed:
Code Works with Aloha Solo,Static

shantanuparab-tr committed:
Aloha Mobile Record and Replay

shantanuparab-tr committed:
Unclean code

shantanuparab-tr committed:
Visualize Dataset Edited

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -71,8 +70,8 @@ def __init__(
         node: InterbotixRobotNode,
```
Unable to comment on this directly, but the observation and action spaces are no longer fixed. We can likely remove the doc comment above for them.

shantanuparab-tr replied:
Should we update the docstring to reflect the customizable nature of these spaces, or would you prefer removing it entirely?

lukeschmitt-tr replied:
Documentation describing the spaces could be kept there or moved elsewhere. I think that being buried this low in the code may make it difficult to find. Any thoughts on where that should go?

shantanuparab-tr replied:
Yes I guess that's true. I'll remove this and add it to the documentation later.

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -71,8 +70,8 @@ def __init__(
         setup_base: bool = False,
-        is_mobile: bool = IS_MOBILE,
         torque_base: bool = False,
+        config: dict = None
```
Modify parameter descriptions below

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -85,33 +84,47 @@ def __init__(
-        self.recorder_right = Recorder('right', node=node)
-        self.image_recorder = ImageRecorder(node=node, is_mobile=IS_MOBILE)
+        self.is_mobile = config.get('base',False)
+        print(f"IS MObile is {self.is_mobile}")
```
Do we need this print statement?

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -85,33 +84,47 @@ def __init__(
+        print(f"IS MObile is {self.is_mobile}")
+        # Dynamically import module based on config value
+        if self.is_mobile:
+            print("Importing Slate")
```
Do we need this?

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -134,35 +147,58 @@ def setup_base(self, node: InterbotixRobotNode, torque_enable: bool = False):
         setup_follower_bot(self.follower_bot_left)
         setup_follower_bot(self.follower_bot_right)
-
+    
```
Recommend using the Trailing Spaces VSCode extension to spot and remove these

```suggestion

```

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -134,35 +147,58 @@ def setup_base(self, node: InterbotixRobotNode, torque_enable: bool = False):
+
+        # Iterate through all follower robots in the self.robots dictionary
+        for name, bot in self.robots.items():
+            if "follower" in name:
```
Make sure we're documenting this requirement somewhere

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -134,35 +147,58 @@ def setup_base(self, node: InterbotixRobotNode, torque_enable: bool = False):
+        for name, bot in self.robots.items():
+            if "follower" in name:
+                # Get the arm joint positions (first 6 joints)
+                arm_qpos = bot.core.joint_states.position[:6]
```
I'm surprised we don't have a bot.arm.get_joint_positions() method. I'll add that to the xs_modules so we can use that instead of reading directly from the joint states and hardcoding the number of joint positions we want to read. This current implementation forces us to use arms with only 6 joints.

lukeschmitt-tr replied:
I have added methods to the arm and gripper modules that can be used to get joint positions: https://github.com/Interbotix/interbotix_ros_toolboxes/tree/xs_robot_get_js

Go ahead and use those here instead of hardcoding the number of joints.

shantanuparab-tr committed:
Added Helper Functions

shantanuparab-tr commented:
## Key Changes
- **Modular YAML Loading**: The `load_yaml_file` function now dynamically loads robot and task configurations, supporting absolute paths and streamlined error handling for robust configuration management. The `load_yaml_file` function is moved to the `utils` file as helper function.
- **Camera Initialization Overhaul**: Camera setup in `ImageRecorder` has been refactored to dynamically use configuration-defined camera names and topics. This reduces hardcoding and simplifies adjustments for different camera setups.
- **FPS-Based Timing Control**: Introduced FPS-driven timing adjustments for consistent durations in motion and observation steps, calculated dynamically from configuration values. This ensures smoother and predictable robot behavior.
- **Episode Index Check with Overwrite Confirmation**: Added a utility to verify episode indices before writing, prompting the user for confirmation if an episode already exists. This change safeguards against accidental overwrites and enhances data integrity.

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -134,35 +146,58 @@ def setup_base(self, node: InterbotixRobotNode, torque_enable: bool = False):
+                qpos_list.append(arm_qpos)
+
+                # Get the gripper joint position (7th joint) and normalize it
+                gripper_qpos = [FOLLOWER_GRIPPER_POSITION_NORMALIZE_FN(bot.core.joint_states.position[7])]
```
Same as above - I'll add a bot.gripper.get_position() command

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -172,28 +207,33 @@ def get_base_vel(self):
-        )
-        self.gripper_command.cmd = right_gripper_desired_joint
-        self.follower_bot_right.gripper.core.pub_single.publish(self.gripper_command)
+        Args:
```
Adhere to the doc comment style in the RealEnv constructor

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -229,9 +276,10 @@ def get_reward(self):
 
     def reset(self, fake=False):
```
Should we just remove the "fake" robot? We don't really support the MuJoCo simulation as of now

shantanuparab-tr replied:
Sure, will do that

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -240,39 +288,49 @@ def reset(self, fake=False):
             reward=self.get_reward(),
             discount=None,
-            observation=obs)
+            observation=obs
```
```suggestion
            observation=obs,
```

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -240,39 +288,49 @@ def reset(self, fake=False):
         )
+    
+    def step(self, action, base_action=None, get_obs=True):
+        follower_bots = {name: robot for name, robot in self.robots.items() if 'follower' in name}
```
We could probably create a self.leaders and self.followers in the constructor for RealEnv - it seems like we're looping through the robots searching for both groups multiple times

shantanuparab-tr replied:
I did that for the followers. But for leaders there is just one instance where they are used and that function is not a member of the class.

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -282,6 +340,7 @@ def make_real_env(
     setup_robots: bool = True,
     setup_base: bool = False,
     torque_base: bool = False,
+    config: dict = None
```
```suggestion
    config: dict = None,
```

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -291,8 +350,8 @@ def make_real_env(
         setup_base=setup_base,
-        is_mobile=IS_MOBILE,
         torque_base=torque_base,
+        config = config
```
```suggestion
        config = config,
```

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -21,154 +18,81 @@
+        # Get camera names from config dictionary
+        self.camera_names = [camera['name'] for camera in config.get('cameras', {}).get('camera_instances', [])]
+
+        COLOR_IMAGE_TOPIC_NAME = config.get('cameras',{}).get('common_parameters', {}).get('color_image_topic_name', None)
```
No longer a const - make snake_case

```suggestion
        color_image_topic_name = config.get('cameras',{}).get('common_parameters', {}).get('color_image_topic_name', None)
```

lukeschmitt-tr replied:
Can grab camera configs first and use it twice where needed. OR we can just pass the cameras dict into this object?

shantanuparab-tr committed:
Removed Unnecessary Passing of Config

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -181,8 +105,10 @@ def get_arm_gripper_positions(bot: InterbotixManipulatorXS):
-    moving_time: float = 1.0,
+    moving_time: float = 1.0
+    
 ) -> None:
```
```suggestion
    moving_time: float = 1.0,
) -> None:
```

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -209,21 +136,24 @@ def sleep_arms(
-            [[0.0, -0.96, 1.16, 0.0, -0.3, 0.0]] * len(bot_list),
+            bot_list=bot_list,
+            DT=DT,
+            target_pose_list=[[0.0, -0.96, 1.16, 0.0, -0.3, 0.0]] * len(bot_list),
```
This list should probably be a config (maybe per robot model?) - we can defer this for now, but we should add a TODO here

shantanuparab-tr committed:
Removed YAML files from Robot Config

shantanuparab-tr committed:
Code clean up and commenting

shantanuparab-tr committed:
Docstring Formatting

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -85,33 +83,48 @@ def __init__(
-        self.recorder_left = Recorder('left', node=node)
-        self.recorder_right = Recorder('right', node=node)
-        self.image_recorder = ImageRecorder(node=node, is_mobile=IS_MOBILE)
+        self.is_mobile = config.get('base',False)
```
Make sure to lint these files

```suggestion
        self.is_mobile = config.get('base', False)
```

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -21,155 +18,83 @@
-        else:
-            self.camera_names = ['cam_high', 'cam_low', 'cam_left_wrist', 'cam_right_wrist']
+        # Get camera names from config dictionary
+        self.camera_names = [camera['name'] for camera in config.get(
```
Can get the cameras_config once and use it twice below to help with readability - something like

```py
cameras_config = config.get('cameras', {})
```

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -21,155 +18,83 @@
+        self.camera_names = [camera['name'] for camera in config.get(
+            'cameras', {}).get('camera_instances', [])]
+
+        COLOR_IMAGE_TOPIC_NAME = config.get('cameras', {}).get(
```
this is no longer a const - make this `color_image_topic_name`

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -181,8 +106,10 @@ def get_arm_gripper_positions(bot: InterbotixManipulatorXS):
 
 def move_arms(
     bot_list: Sequence[InterbotixManipulatorXS],
+    DT: float,
```
this is no longer a const - make this `dt`

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -198,8 +125,9 @@ def move_arms(
+    DT: float,
     moving_time: float = 5.0,
-    home_first: bool = True,
+    home_first: bool = True
```
```suggestion
    home_first: bool = True,
```

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -181,8 +106,10 @@ def get_arm_gripper_positions(bot: InterbotixManipulatorXS):
-    moving_time: float = 1.0,
+    moving_time: float = 1.0
+
 ) -> None:
```
```suggestion
    moving_time: float = 1.0,
) -> None:
```

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -209,21 +137,26 @@ def sleep_arms(
     if home_first:
         move_arms(
-            bot_list,
-            [[0.0, -0.96, 1.16, 0.0, -0.3, 0.0]] * len(bot_list),
```
We should probably make this a parameter somewhere - not important to do now, but you could add a TODO. it's also the first 6 elements of the START_ARM_POSE constant

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -209,21 +137,26 @@ def sleep_arms(
+        target_pose_list=[
+            bot.arm.group_info.joint_sleep_positions for bot in bot_list],
         moving_time=moving_time,
+        DT=DT
```
```suggestion
        DT=DT,
```

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -209,21 +137,26 @@ def sleep_arms(
     bot_list: Sequence[InterbotixManipulatorXS],
     target_pose_list: Sequence[float],
     moving_time: float,
+    DT
```
Make sure to add a comma even to the last parameter so that changes to the signature look clean later

```suggestion
    DT,
```

lukeschmitt-tr commented on config/robot/aloha_mobile.yaml:
```diff
@@ -0,0 +1,59 @@
+robot:
+  # Basic robot configuration
+  name: aloha_mobile                # Unique identifier for the robot
+  fps: 50                         # Frames per second for control and data updates
```
Put the parameter comments above the parameters they describe, include defaults, if any

lukeschmitt-tr commented on config/robot/aloha_remote.yaml:
Let's only provide solo, static (stationary?), and mobile

shantanuparab-tr replied:
Sure, I'll change static to stationary and remove others.

lukeschmitt-tr commented on config/rs_cam.yaml:
```diff
@@ -16,7 +16,7 @@
   camera:
     ros__parameters:
-      serial_no: ""
+      serial_no: "128422271347"
```
Revert the changes to this file

shantanuparab-tr replied:
I don't see anymore use of the rs_cam.yaml file. Should we delete it?

lukeschmitt-tr replied:
Good point! We can remove this file, yeah.

lukeschmitt-tr commented on config/tasks_config.yaml:
```diff
@@ -0,0 +1,23 @@
+
+# Real hardware task configurations
+tasks:
+  aloha_mobile_hello_aloha:
```
Let's only provide the dummy tasks

lukeschmitt-tr commented on launch/aloha_bringup.launch.py:
```diff
@@ -13,596 +15,424 @@
     PathJoinSubstitution,
+    FindExecutable,
+    Command,
+    TextSubstitution,
```
Alphabetize

lukeschmitt-tr commented on launch/aloha_bringup.launch.py:
```diff
@@ -13,596 +15,424 @@
+            "robot",
+            TextSubstitution(text=f"{robot_config_base}.yaml"),
+        ]
+    ).perform(context)
```
Can do this to help with readability, reduce indentation

```suggestion
    yaml_file_path = PathJoinSubstitution([
        FindPackageShare("aloha"),
        "config",
        "robot",
        TextSubstitution(text=f"{robot_config_base}.yaml"),
    ]).perform(context)
```

lukeschmitt-tr commented on launch/aloha_bringup.launch.py:
```diff
@@ -13,596 +15,424 @@
-        condition=IfCondition(LaunchConfiguration('launch_leaders')),
-    )
+    # Retrieve and display the robot_name
+    robot_name = config.get("name", "Unnamed Robot")
```
Let's call it "aloha" :)

lukeschmitt-tr commented on launch/aloha_bringup.launch.py:
```diff
@@ -13,596 +15,424 @@
-    )
+    # Retrieve and display the robot_name
+    robot_name = config.get("name", "Unnamed Robot")
+    print(f"Robot Name: {robot_name}")
```
Debug statement?

lukeschmitt-tr commented on launch/aloha_bringup.launch.py:
```diff
@@ -13,596 +15,424 @@
-        output={'both': 'log'},
-    )
+        xsarm_control_leader_launch_include = IncludeLaunchDescription(
+            PythonLaunchDescriptionSource(
```
Apply the same suggestion as above regarding indentation and readability

shantanuparab-tr replied:
What's the correct indentation format?

lukeschmitt-tr commented on launch/aloha_bringup.launch.py:
```diff
@@ -13,596 +15,424 @@
-        ]),
-    )
+    # Check if base is enabled in the YAML configuration
+    if base_enabled:
```
This variable is only used here - get the parameter here instead of creating the variable

lukeschmitt-tr commented on launch/aloha_bringup.launch.py:
```diff
@@ -13,596 +15,424 @@
-        emulate_tty=True,
-        parameters=[{'motor_specs': LaunchConfiguration('leader_motor_specs_right')}],
-        condition=IfCondition(LaunchConfiguration('use_gravity_compensation')),
+    loginfo_action = LogInfo(
```
This can probably be changed a bit to reflect our new approach

lukeschmitt-tr commented on launch/aloha_bringup.launch.py:
```diff
@@ -13,596 +15,424 @@
 
 def generate_launch_description():
+
+    print("Generating Launch Description")
```
Debugging statement?

lukeschmitt-tr reviewed (changes requested):
Took a pass at half the files here

shantanuparab-tr committed:
Reviewed Code Fix

shantanuparab-tr committed:
Reviewed Code Fix

lukeschmitt-tr commented on .gitignore:
```diff
@@ -7,3 +7,6 @@ log*
 act/
 data/
+config/robot/aloha_remote.yaml
+config/robot/aloha_shoot.yaml
```
Just don't commit these files or move them outside of the git repo instead of adding them to the global gitignore

lukeschmitt-tr commented on config/robot/aloha_solo.yaml:
```diff
@@ -0,0 +1,86 @@
+        name: camera_wrist_right             
+
+        # Serial number for device identification
+        serial_no: "130322272300"            
```
Leave all serial numbers unset

lukeschmitt-tr commented on scripts/visualize_episodes.py:
```diff
@@ -1,22 +1,21 @@
 
 
-def load_hdf5(dataset_dir, dataset_name):
+def load_hdf5(dataset_dir, dataset_name, IS_MOBILE):
```
snake_case the IS_MOBILE variable and all other non-constants

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -307,3 +253,132 @@ def enable_gravity_compensation(bot: InterbotixManipulatorXS):
+    """
+
+    # Base path of the config directory using absolute path
+    base_path = os.path.abspath(os.path.join("..", "config"))
```
This seems like a fragile way to do this. Should we pass in the absolute filepath instead?

shantanuparab-tr replied:
Used pathlib for handling path. Now the python script can be executed from any location.

lukeschmitt-tr commented on config/robot/aloha_mobile.yaml:
```diff
@@ -0,0 +1,114 @@
+      # Model type for compatibility and calibration
+      model: aloha_wx250s              
+
+      # Orientation side (options: left, right)
+      orientation: right               
```
There's probably a better way to do this than setting the orientation. I know this is used to load modes files, right? Can we accomplish this a better way? Can we create directories per robot type and store the configs in here? Like for the solo, it doesn't make sense to tell the users to create a symlink to ttyDXL_leader_left

lukeschmitt-tr commented on config/robot/aloha_solo.yaml:
```diff
@@ -0,0 +1,86 @@
+      orientation: right           
+
+      # Transformation matrix for spatial alignment (used in RVIZ or other visualization tools)
+      transform: [-0.5, -0.25, 0.0, 0.0, 0.0, 0.0, 1.0]  
```
```suggestion
      transform: [-0.5, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]  
```

lukeschmitt-tr commented on config/robot/aloha_solo.yaml:
```diff
@@ -0,0 +1,86 @@
+      orientation: right           
+
+      # Transformation matrix for spatial alignment
+      transform: [0.5, -0.25, 0.0, 0.0, 0.0, 0.0, 1.0]  
```
```suggestion
      transform: [0.5, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]  
```

lukeschmitt-tr commented on config/rs_cam.yaml:
Aren't we removing this file now?

shantanuparab-tr commented on scripts/record_episodes.py:
```diff
@@ -40,379 +27,522 @@
 
-    # Data collection
+    # Begin data collection
     ts = env.reset(fake=True)
```
Reset Env fake is required while recording episodes. So, I'll  changing the reset function.

shantanuparab-tr commented on aloha/real_env.py:
```diff
@@ -203,85 +220,109 @@ def _reset_gripper(self):
+                robot.core.robot_reboot_motors('single', 'gripper', True)
+        self._reset_joints()
+        self._reset_gripper()
         return dm_env.TimeStep(
```
Had to add fake to handle record episode script.

shantanuparab-tr committed:
Fixed Reviewed Code: Part 1

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -1,122 +1,113 @@
+    FOLLOWER_GRIPPER_POSITION_NORMALIZE_FN,
+    FOLLOWER_GRIPPER_VELOCITY_NORMALIZE_FN,
+    LEADER_GRIPPER_JOINT_NORMALIZE_FN,
+    START_ARM_POSE
```
```suggestion
    START_ARM_POSE,
```

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -1,122 +1,113 @@
+                iterative_update_fk=False,
+            )
+
+        self.follower_bots = [robot for name,
+                              robot in self.robots.items() if 'follower' in name]
+
+        self.follower_robots = {name: robot for name,
+                                robot in self.robots.items() if 'follower' in name}
```
dict and list?

shantanuparab-tr replied:
I am trying to figure out a better solution for this.

shantanuparab-tr replied:
I have shifed it to the method where it is required. But I guess there is a more elegant way to handle it.

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -1,122 +1,113 @@
         setup_base: bool = False,
-        is_mobile: bool = IS_MOBILE,
         torque_base: bool = False,
+        config: dict = None
```
Make sure you're adding commas to the end of all multi-line method param declarations

```suggestion
        config: dict = None,
```

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -1,122 +1,113 @@
+        """
+        self.is_mobile = config.get('base', False)
+
+        self.DT = 1 / config.get('fps', 30)
```
Not const

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -1,122 +1,113 @@
+                node=node,
+                iterative_update_fk=False,
+            )
+
```
Should we raise if `robots` is empty?

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -136,33 +127,56 @@ def setup_robots(self):
+        qpos_list = []
+
+        # Iterate through all follower robots in the self.robots dictionary
+        for name, bot in self.robots.items():
```
We should document that the order of the observations relies on the order of the arms. We should also make sure that we're backwards compatible with the previous version of aloha - that we're saving the observations in the same order as before for all robot styles.

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -136,33 +127,56 @@ def setup_robots(self):
         setup_follower_bot(self.follower_bot_right)
 
     def get_qpos(self):
```
Do we want to go through the effort of adding return typing and docstrings to each method?

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -172,29 +186,35 @@ def get_base_vel(self):
+            # Repeat reset_position for each robot
+            target_pose_list=[reset_position] * len(self.follower_bots),
             moving_time=1.0,
+            dt=self.DT
```
```suggestion
            dt=self.DT,
```

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -241,38 +270,52 @@ def reset(self, fake=False):
+        for name, robot in self.follower_robots.items():
+            bot_action = action[index:index + state_len]
+            robot.arm.set_joint_positions(
+                bot_action[:6], blocking=False)  # Set arm positions
```
Allows us to use arms of different size assuming that actions look like [joint_0, joint_1, ..., joint_n, gripper]

```suggestion
                bot_action[:-1], blocking=False)  # Set arm positions
```

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -241,38 +270,52 @@ def reset(self, fake=False):
-    # Gripper actions
-    action[6] = LEADER_GRIPPER_JOINT_NORMALIZE_FN(leader_bot_left.core.joint_states.position[6])
-    action[7+6] = LEADER_GRIPPER_JOINT_NORMALIZE_FN(leader_bot_right.core.joint_states.position[6])
+def get_action(robots: dict):
```
More precise typing of dicts

```suggestion
def get_action(robots: dict[str, InterbotixManipulatorXS]):
```

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -241,38 +270,52 @@ def reset(self, fake=False):
+
+    num_leader_bots = len(leader_bots)
+    # 6 joint positions + 1 gripper for each leader bot
+    action = np.zeros(num_leader_bots * 7)
```
Can use bot.arm.group_info.num_joints to get the number of joints in the arm group

lukeschmitt-tr commented on aloha/real_env.py:
```diff
@@ -241,38 +270,52 @@ def reset(self, fake=False):
+    action = np.zeros(num_leader_bots * 7)
+
+    index = 0
+    for robot_name, robot in leader_bots.items():
```
If you just want the robots, you can use leader_bots.values() instead of items(). I think you do this in a few places.

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -1,175 +1,102 @@
+        self.camera_names = [camera['name']
+                             for camera in camera_config.get('camera_instances', [])]
+
+        color_image_topic_name = config.get('cameras', {}).get(
```
```suggestion
        color_image_topic_name = camera_config.get(
```

lukeschmitt-tr commented on aloha/robot_utils.py:
```diff
@@ -1,175 +1,102 @@
+                             for camera in camera_config.get('camera_instances', [])]
+
+        color_image_topic_name = config.get('cameras', {}).get(
+            'common_parameters', {}).get('color_image_topic_name', None)
```
Should we assert or throw that these configs have been set to not `None`?

shantanuparab-tr replied:
Will do that.

shantanuparab-tr commented on aloha/real_env.py:
```diff
@@ -136,33 +127,56 @@ def setup_robots(self):
+
+                # Get the gripper joint velocity (7th joint) and normalize it
+                gripper_qvel = [FOLLOWER_GRIPPER_VELOCITY_NORMALIZE_FN(
+                    bot.core.joint_states.velocity[6])]
```
We need method to get gripper velocity.

lukeschmitt-tr replied:
https://github.com/Interbotix/interbotix_ros_toolboxes/pull/80/commits/d06005312c1db8cbb956ca05ecfb75f159fc19d4

shantanuparab-tr commented on aloha/real_env.py:
```diff
@@ -136,33 +127,56 @@ def setup_robots(self):
+        for name, bot in self.robots.items():
+            if "follower" in name:
+                # Get the first 7 effort values (6 for arm, 1 for gripper)
+                robot_effort = bot.core.joint_states.effort[:7]
```
We also need method to get gripper effort. I'll concatenate the arm and gripper efforts here.

lukeschmitt-tr replied:
Added in https://github.com/Interbotix/interbotix_ros_toolboxes/pull/80/commits/d06005312c1db8cbb956ca05ecfb75f159fc19d4

lukeschmitt-tr commented on scripts/sleep.py:
```diff
@@ -1,11 +1,12 @@
     sleep_arms,
     torque_on,
     disable_gravity_compensation,
+    load_yaml_file
```
```suggestion
    load_yaml_file,
```

lukeschmitt-tr commented on scripts/visualize_episodes.py:
```diff
@@ -136,7 +147,8 @@ def visualize_joints(qpos_list, command_list, plot_path=None, ylim=None, label_o
     # plot joint state
     all_names = (
-        [f'{name}_left' for name in STATE_NAMES] + [f'{name}_right' for name in STATE_NAMES]
+        [f'{name}_left' for name in STATE_NAMES] +
```
Does this work for solo?

shantanuparab-tr replied:
Will fix this.

shantanuparab-tr commented on aloha/real_env.py:
```diff
@@ -172,29 +186,35 @@ def get_base_vel(self):
+            self.gripper_command)
 
     def _reset_joints(self):
         reset_position = START_ARM_POSE[:6]
```
We are still using hardcoded values for constants is there a better way?

lukeschmitt-tr replied:
I think it'd make sense to have this be a parameter, but we can get that on a second pass

shantanuparab-tr commented on scripts/record_episodes.py:
```diff
@@ -40,379 +27,522 @@
+                if follower_name in follower_bots:
+                    pairs.append((leader_bot, follower_bots[follower_name]))
+                break
+        else:
```
Removed the hard coded suffixes to have more flexibility in naming.

shantanuparab-tr committed:
Fixed Short Term Requirements

shantanuparab-tr committed:
Renaming and Removing Files

shantanuparab-tr committed:
Fixed load_yaml and no camera setup, removed debugging statements

shantanuparab-tr committed:
Fixed Auto Record and Visualize
